File: sdks/cpp/tests/trends_test.cpp
// Tests for the endurance-tests suite's pass/fail semantics
// (../endurance-tests/trends.cpp). Pure functions, no FFI/reactor::sdk
// involved — mirrors sdks/python/tests/test_trends.py.
#include "trends.hpp"

#include <array>

#include <catch2/catch_approx.hpp>
#include <catch2/catch_test_macros.hpp>

using endurance::assert_always_zero;
using endurance::assert_never_grows;
using endurance::assert_no_sustained_growth;
using endurance::cpu_deltas;
using endurance::Sample;

namespace {

Sample make_sample(int cycle, double cpu_s = 0.0, long num_threads = 20, long num_fds = 15,
                   double cpu_percent = 5.0) {
  Sample s;
  s.cycle = cycle;
  s.elapsed_s = static_cast<double>(cycle);
  s.rss_bytes = 100'000'000;
  s.cpu_s = cpu_s;
  s.cpu_percent = cpu_percent;
  s.num_threads = num_threads;
  s.num_fds = num_fds;
  return s;
}

}  // namespace

TEST_CASE("cpu_deltas diffs consecutive cumulative values") {
  std::vector<Sample> samples;
  samples.reserve(4);
  const std::array<double, 4> values = {0.0, 0.5, 1.2, 1.4};
  for (int i = 0; i < 4; ++i) {
    samples.push_back(make_sample(i, values.at(static_cast<std::size_t>(i))));
  }
  const auto deltas = cpu_deltas(samples);
  REQUIRE(deltas.size() == 3);
  CHECK(deltas[0] == Catch::Approx(0.5));
  CHECK(deltas[1] == Catch::Approx(0.7));
  CHECK(deltas[2] == Catch::Approx(0.2));
}

TEST_CASE("cpu_deltas on empty or single sample yields no deltas") {
  CHECK(cpu_deltas({}).empty());
  CHECK(cpu_deltas({make_sample(0)}).empty());
}

TEST_CASE("assert_no_sustained_growth: too few samples throws, not just fails") {
  // n < 6 is a test misconfiguration (ENDURANCE_DURATION_SECONDS too short),
  // not a metric to report on — see the function's own doc-comment for why
  // this one case still throws directly.
  CHECK_THROWS_AS(assert_no_sustained_growth({1.0, 2.0, 3.0}, "x", 0.15), std::runtime_error);
}

TEST_CASE("assert_no_sustained_growth: flat values pass") {
  const std::vector<double> values(30, 100.0);
  const auto result = assert_no_sustained_growth(values, "rss", 0.15, 5.0, 0.2, false, "MB");
  CHECK(result.status == "ok");
}

TEST_CASE("assert_no_sustained_growth: still climbing in the tail fails") {
  // Genuinely still trending up in the last third relative to the middle
  // third, well past both the ratio and absolute floors.
  std::vector<double> values;
  values.reserve(30);
  for (int i = 0; i < 30; ++i) {
    values.push_back(100.0 + static_cast<double>(i) * 5.0);
  }
  const auto result = assert_no_sustained_growth(values, "rss", 0.15, 5.0, 0.2, false, "MB");
  CHECK(result.status == "fail");
  CHECK(result.detail.rfind("still climbing", 0) == 0);
}

TEST_CASE("assert_no_sustained_growth: a one-time ramp that has already plateaued passes") {
  // The exact shape that motivated middle-vs-last over first-vs-last: flat,
  // then a one-time ramp to a new plateau early in the run, then flat for
  // the rest. The *overall* first-to-last movement is large, but the tail
  // (last third vs middle third) is flat — a native buffer/pool growing
  // once to steady-state, not an unbounded leak, so this must pass.
  constexpr int n = 30;
  std::vector<double> values;
  for (int i = 0; i < n; ++i) {
    const double pct = static_cast<double>(i) / static_cast<double>(n - 1);
    if (pct < 0.25) {
      values.push_back(100.0);
    } else if (pct < 0.45) {
      values.push_back(100.0 + (pct - 0.25) / 0.20 * 40.0);
    } else {
      values.push_back(140.0);
    }
  }
  const auto result = assert_no_sustained_growth(values, "rss", 0.15, 5.0, 0.2, false, "MB");
  CHECK(result.status == "ok");
  // The reported start/end still show the true overall movement, even
  // though that's not what decided pass/fail.
  CHECK(result.change > 20.0);
}

TEST_CASE("assert_no_sustained_growth: min_absolute_delta floor suppresses a tiny-baseline ratio") {
  // A tiny near-zero baseline can turn an insignificant wobble into a ratio
  // that looks huge — the absolute floor exists for exactly this.
  std::vector<double> values(15, 0.01);
  values.insert(values.end(), 15, 0.02);
  const auto result =
      assert_no_sustained_growth(values, "cpu_s_per_cycle", 0.15, 0.05, 0.2, false, "s");
  CHECK(result.status == "ok");
}

TEST_CASE("assert_no_sustained_growth: use_median ignores a single outlier cycle") {
  // A one-cycle double-counted-thread/fd artifact landing in the last
  // window shouldn't by itself read as a sustained trend.
  std::vector<double> values(12, 25.0);
  const std::array<double, 6> tail = {25.0, 25.0, 31.0, 25.0, 25.0, 25.0};
  values.insert(values.end(), tail.begin(), tail.end());
  const auto result = assert_no_sustained_growth(values, "num_threads", 0.15, 4.0, 0.2,
                                                 /*use_median=*/true, "count");
  CHECK(result.status == "ok");
}

TEST_CASE("assert_always_zero: stays zero passes") {
  std::vector<Sample> samples;
  samples.reserve(10);
  for (int i = 0; i < 10; ++i) {
    samples.push_back(make_sample(i));
  }
  const auto result = assert_always_zero(samples, [](const Sample&) { return 0L; }, "orphaned");
  CHECK(result.status == "ok");
}

TEST_CASE("assert_always_zero: nonzero mid-run fails even if it settles back to zero") {
  // Regression: a leak on cycle 3 that happens to get cleaned up by the
  // final cycle is still a real bug — must not read as "ok" just because
  // the last sample is 0.
  const std::array<long, 6> values = {0, 0, 0, 2, 0, 0};
  std::vector<Sample> samples;
  samples.reserve(6);
  for (int i = 0; i < 6; ++i) {
    samples.push_back(make_sample(i));
  }
  const auto result = assert_always_zero(
      samples, [&](const Sample& s) { return values.at(static_cast<std::size_t>(s.cycle)); },
      "orphaned");
  CHECK(result.status == "fail");
  REQUIRE(result.first_bad_cycle.has_value());
  CHECK(result.first_bad_cycle.value() == 3);  // NOLINT(bugprone-unchecked-optional-access)
}

TEST_CASE("assert_always_zero: reports the peak, not the final sample") {
  // Regression (codex review on PR #186 upstream, ported here): using the
  // *final* sample's value for end/change would render "went from 0 to 0
  // (+0)" whenever the count had already settled back to 0 by the last
  // cycle — hiding the leaked value entirely.
  const std::array<long, 6> values = {0, 0, 5, 3, 0, 0};
  std::vector<Sample> samples;
  samples.reserve(6);
  for (int i = 0; i < 6; ++i) {
    samples.push_back(make_sample(i));
  }
  const auto result = assert_always_zero(
      samples, [&](const Sample& s) { return values.at(static_cast<std::size_t>(s.cycle)); },
      "orphaned");
  CHECK(result.status == "fail");
  CHECK(result.end == 5);
  CHECK(result.change == 5);
  CHECK(result.detail.find("peak 5") != std::string::npos);
}

TEST_CASE("assert_never_grows: flat at baseline passes") {
  std::vector<Sample> samples;
  samples.reserve(10);
  for (int i = 0; i < 10; ++i) {
    samples.push_back(make_sample(i));
  }
  const auto result =
      assert_never_grows(samples, [](const Sample& s) { return s.num_threads; }, "num_threads");
  CHECK(result.status == "ok");
}

TEST_CASE("assert_never_grows: growth past the starting value fails") {
  const std::array<long, 5> values = {1, 1, 1, 2, 1};
  std::vector<Sample> samples;
  samples.reserve(5);
  for (int i = 0; i < 5; ++i) {
    samples.push_back(make_sample(i));
  }
  const auto result = assert_never_grows(
      samples, [&](const Sample& s) { return values.at(static_cast<std::size_t>(s.cycle)); },
      "live_clients");
  CHECK(result.status == "fail");
  CHECK(result.start == 1);
  CHECK(result.end == 2);
}

TEST_CASE("standard_resource_metrics returns the shared five checks") {
  std::vector<Sample> samples;
  samples.reserve(30);
  for (int i = 0; i < 30; ++i) {
    samples.push_back(make_sample(i, /*cpu_s=*/static_cast<double>(i) * 0.01));
  }
  const auto metrics = endurance::standard_resource_metrics(samples);
  // rss, cpu_s_per_cycle, cpu_percent, num_threads, num_fds — no
  // live_clients/orphaned_callbacks equivalent in this binding (see
  // ../../endurance-tests/README.md).
  REQUIRE(metrics.size() == 5);
  for (const auto& m : metrics) {
    CHECK(m.status == "ok");
  }
}

File: sdks/android/endurance-tests/src/test/kotlin/inc/reactor/sdk/android/endurance/TrendsTest.kt
package inc.reactor.sdk.android.endurance

import org.junit.Assert.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Assert.assertTrue
import org.junit.Test

/**
 * The verdicts, against runs whose shape is known because this file made it.
 *
 * These exist because the thing they check cannot be checked any other way: a trend rule is only
 * wrong on a shape you did not happen to run, and the real suite takes minutes on hardware to
 * produce one shape. Here a leak, a plateau and a flat line are three lines of arithmetic each.
 */
class TrendsTest {
    private fun run(values: List<Double>): List<Sample> =
        values.mapIndexed { index, value ->
            Sample(
                cycle = index + 1,
                elapsedSeconds = index.toDouble(),
                rssBytes = (value * 1_048_576).toLong(),
                cpuNanos = 0,
                threads = 10,
                openFds = 20,
                liveClients = 0,
                orphanedGlobals = 0,
                nativeHeapBytes = 0,
            )
        }

    private fun rss(samples: List<Sample>) =
        Trends.assertNoSustainedGrowth(samples, "rss", " MB", 0.10, 8.0, false) {
            it.rssBytes / 1_048_576.0
        }

    @Test
    fun `a flat run passes`() {
        val result = rss(run(List(60) { 100.0 }))
        assertTrue(result.detail, result.passed)
        assertEquals(0.0, result.midToEnd, 0.001)
    }

    @Test
    fun `a steadily climbing run fails`() {
        val result = rss(run(List(60) { 100.0 + it * 5 }))
        assertFalse(result.detail, result.passed)
        assertTrue(result.detail.contains("still climbing"))
    }

    /**
     * The case the mid-third comparison exists for.
     *
     * A one-time ramp early on, then flat. Against the *first* third this reads as large growth
     * and fails; against the middle third it reads as what it is.
     */
    @Test
    fun `a one-time ramp that plateaus passes`() {
        val values = List(20) { 100.0 } + List(10) { 100.0 + it * 20.0 } + List(30) { 300.0 }
        val result = rss(run(values))
        assertTrue(result.detail, result.passed)
        assertTrue("the report must still show the step from start", result.end > result.start)
    }

    /** Growth under the absolute floor is noise, whatever the ratio says. */
    @Test
    fun `a tiny absolute rise passes even though the ratio is large`() {
        val values = List(30) { 1.0 } + List(30) { 4.0 }
        val result = rss(run(values))
        assertTrue(result.detail, result.passed)
    }

    @Test
    fun `too few samples is reported rather than guessed at`() {
        val result = rss(run(List(4) { 100.0 }))
        assertTrue(result.passed)
        assertTrue(result.detail, result.detail.contains("too few samples"))
    }

    // ── assertAlwaysZero ─────────────────────────────────────────────────────

    @Test
    fun `a value that is zero throughout passes`() {
        val result =
            Trends.assertAlwaysZero(run(List(30) { 1.0 }), "orphaned globals") {
                it.orphanedGlobals.toDouble()
            }
        assertTrue(result.passed)
    }

    /**
     * The reason this check exists rather than a trend check: a leak present from cycle 0 is
     * perfectly flat, and a trend check would accept it forever.
     */
    @Test
    fun `a value already nonzero on the first cycle fails, where a trend check would not`() {
        val samples = run(List(30) { 1.0 }).map { it.copy(orphanedGlobals = 3) }
        val zero = Trends.assertAlwaysZero(samples, "orphaned globals") { it.orphanedGlobals.toDouble() }
        assertFalse(zero.passed)
        assertTrue(zero.detail, zero.detail.contains("cycle 1"))

        val trend =
            Trends.assertNoSustainedGrowth(samples, "orphaned globals", "", 0.10, 0.0, true) {
                it.orphanedGlobals.toDouble()
            }
        assertTrue("a flat nonzero line is exactly what a trend check accepts", trend.passed)
    }

    @Test
    fun `a value that leaves zero midway names the cycle it happened on`() {
        val samples =
            run(List(30) { 1.0 }).mapIndexed { index, sample ->
                if (index >= 20) sample.copy(orphanedGlobals = 1) else sample
            }
        val result = Trends.assertAlwaysZero(samples, "orphaned globals") { it.orphanedGlobals.toDouble() }
        assertFalse(result.passed)
        assertTrue(result.detail, result.detail.contains("cycle 21"))
    }

    // ── the standard set ─────────────────────────────────────────────────────

    /**
     * fds get the trend check, never an exact never-grows.
     *
     * Two bindings independently caught a one-cycle step that settled back down before the run
     * ended — native socket teardown landing mid-sample, not a leak. A never-grows rule fails that
     * run; the median-backed trend does not.
     */
    @Test
    fun `a single fd spike that settles back does not fail the run`() {
        val samples =
            run(List(60) { 100.0 }).mapIndexed { index, sample ->
                sample.copy(openFds = if (index == 42) 60 else 20)
            }
        val fds = Trends.standardResourceMetrics(samples, false).first { it.name == "open fds" }
        assertTrue(fds.detail, fds.passed)
    }

    @Test
    fun `live clients is only checked when the scenario returns to zero`() {
        val samples = run(List(30) { 1.0 }).map { it.copy(liveClients = 1) }
        assertTrue(Trends.standardResourceMetrics(samples, false).none { it.name == "live clients" })
        val checked = Trends.standardResourceMetrics(samples, true).first { it.name == "live clients" }
        assertFalse("a churn scenario holding a client open is a failure", checked.passed)
    }
}
